lib/tree.rb

```diff
@@ -1,62 +1,238 @@
+    end
+    return current_node
+  end
+
   def add(key, value)
-    raise NotImplementedError
+    @root = add_helper(@root, key, value)
+  end
+
+  # Time Complexity: O(log n)
+  # Space Complexity: O(n)
+  def find_helper(current_node, key)
+    return nil if current_node.nil?
+    return current_node.value if current_node.key == key
+
+    if key < current_node.key
+      current_node = find_helper(current_node.left, key)
+    else
+      current_node = find_helper(current_node.right, key)
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
   def find(key)
-    raise NotImplementedError
+    find_helper(@root, key)
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
+  def inorder_helper(current_node, list)
+    return list if current_node.nil?
+
+    # left side
+    inorder_helper(current_node.left, list)
+
+    # middle node
+    list << { key: current_node.key, value: current_node.value }
+
+    # right side
+    inorder_helper(current_node.right, list)
+
+    return list
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
   def inorder
-    raise NotImplementedError
+    return inorder_helper(@root, [])
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
+  def preorder_helper(current_node, list)
+    return list if current_node.nil?
+
+    # middle node
+    list << { key: current_node.key, value: current_node.value }
+
+    # left side
+    preorder_helper(current_node.left, list)
+
+    # right side
+    preorder_helper(current_node.right, list)
+
+    return list
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
   def preorder
-    raise NotImplementedError
+    return preorder_helper(@root, [])
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
+  def postorder_helper(current_node, list)
+    return list if current_node.nil?
+
+    # left side
+    postorder_helper(current_node.left, list)
+
+    # right side
+    postorder_helper(current_node.right, list)
+
+    # middle node
+    list << { key: current_node.key, value: current_node.value }
+
+    return list
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
   def postorder
-    raise NotImplementedError
+    return postorder_helper(@root, [])
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(h) - h is height of tree
+  def height_helper(current_node)
+    if current_node.nil?
+      return 0
+    else
+      left_height = height_helper(current_node.left)
+      right_height = height_helper(current_node.right)
+
+      if left_height > right_height
+        return left_height + 1
+      else
+        return right_height + 1
+      end
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
   def height
-    raise NotImplementedError
+    return height_helper(@root)
   end
 
-  # Optional Method
-  # Time Complexity: 
-  # Space Complexity: 
+  # Optional Methods
+  # Time Complexity: O(n)
+  # Space Complexity: O(w) - w is width of tree
   def bfs
```

Comment:
The space complexity of `bfs` in this instance is O(n) since you're building an array.  If you were just printing... then you would be correct, although if the tree is balanced w would be related to the number of nodes.